src/jit/codegenxarch.cpp

```diff
@@ -6252,6 +6252,51 @@ void CodeGen::genCallInstruction(GenTreePtr node)
 
     if (target != nullptr)
     {
+#ifdef _TARGET_X86_
+        if (((call->gtFlags & GTF_CALL_VIRT_KIND_MASK) == GTF_CALL_VIRT_STUB) && (call->gtCallType == CT_INDIRECT))
+        {
+            // On x86, we need to generate a very specific pattern for indirect VSD calls:
+            //
+            //    3-byte nop
+            //    call dword ptr [eax]
+            //
+            // Where EAX is also used as an argument to the stub dispatch helper. Make
+            // sure that the call target address is computed into EAX in this case.
+
+            assert(target->isContainedIndir());
+
+            // Disable random NOP emission
+            getEmitter()->emitDisableRandomNops();
+
+            GenTreeIndir* indir = target->AsIndir();
+            assert(indir->Addr() == indir->Base());
+            assert(indir->HasBase());
+            assert(!indir->HasIndex());
+            assert(indir->Scale() == 1);
+            assert(indir->Offset() == 0);
```

Comment:
We don't need to constrain or make assumptions on what Addr should look like in codegen.  It could be any valid addr mode. This provides flexibility to on what VSD stub addr should look like if it ever changes in future.
Plus, we have utility routines to consume addrs.  It is best to use those utility routines for consistency and uniformity.

---
In reply to: [79260126](https://github.com/dotnet/coreclr/pull/7232#discussion_r79260126) [](ancestors = 79260126)